Add improvedigitalBidAdapter

agregorio-improve opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other

## Description of change
<!-- Describe the change proposed in this pull request -->

<!-- For new bidder adapters, please provide the following -->
- test parameters for validating bids
```
{
  bidder: 'improvedigital',
  params: {
    placementId:1053688
  }
},
{
  bidder: 'improvedigital',
  params: {
    placementId:1053689,
    keyValues: {
      testKey: ["testValue"]
    }
},
{
  bidder: 'improvedigital',
  params: {
    placementId:1053687,
    size: {
      w:300,
      h:300
    }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer: hb@improvedigital.com
- [x] official adapter submission


## Other information
<!-- References to related PR or issue #s, @mentions of the person or team responsible for reviewing changes, etc. -->

harpere commented:
@agregorio-improve The adapter needs unit tests.  You can see examples in `test/spec/modules/`

derdeka commented:
The length of the bidder name "improvedigital" is to long when using enableSendAllBids() with dfp (hb_size_improvedigital, hb_adid_improvedigital, hb_pb_improvedigital). The limit of dfp's key length is 20. Please use a shorter name or register an alias.

derdeka commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,415 @@
+};
+
+ImproveDigitalAdapter.createNew = function () {
+    return new BeachfrontAdapter();
```
Where is BeachfrontAdapter defined? This should be ImproveDigitalAdapter, right?

agregorio-improve replied:
This has been fixed to ImproveDigitalAdapter.

harpere commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,415 @@
+
+var ImproveDigitalAdapter = function ImproveDigitalAdapter() {
+  var IMPROVE_DIGITAL_BIDDER_CODE = "improvedigital";
+  var baseAdapter = Adapter.createNew(IMPROVE_DIGITAL_BIDDER_CODE);
```
`Adapter.createNew` no longer exists.  this should be `var baseAdapter = new Adapter(IMPROVE_DIGITAL_BIDDER_CODE);`

agregorio-improve replied:
This has now been fixed in the latest version.

harpere commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,415 @@
+    getNormalizedBidRequest: baseAdapter.getNormalizedBidRequest,
+    callBids: baseAdapter.callBids
+  };
+};
```
this will need change to (see https://github.com/prebid/Prebid.js/pull/1459)
```
return Object.assign(this, {
    LIB_VERSION: LIB_VERSION,
    idClient: baseAdapter.idClient,
    getNormalizedBidRequest: baseAdapter.getNormalizedBidRequest,
    callBids: baseAdapter.callBids
  });
```

agregorio-improve replied:
This has also been fixed in the latest version.

agregorio-improve commented:
I have now added unit tests and made fixes as per your comments.

I have not set up an alias because it seems that the latest implementation of enableSendAllBids seems to truncate the keys to 20 characters.  Is aliasing still required?

agregorio-improve commented:
Hello,

Can you please take a look at and approve the latest changes?

Best Regards,
Alfredo

muuki88 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,408 @@
+var LIB_VERSION_GLOBAL = '3.0.2';
```
Why not const?

agregorio-improve replied:
Fixed

muuki88 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,408 @@
+var adaptermanager = require('src/adaptermanager');
+
+const ImproveDigitalAdapter = function () {
+  var IMPROVE_DIGITAL_BIDDER_CODE = 'improvedigital';
```
I'll use `const` where ever possible. And `let` instead of `var` for the rest.

muuki88 replied:
The bidder code is also used in the `createNew` function. Maybe extracting this into a `const` would be a good idea.

muuki88 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,408 @@
+      singleRequestMode: false,
+      httpRequestType: this.idClient.CONSTANTS.HTTP_REQUEST_TYPE.GET,
+      callback: CALLBACK_FUNCTION,
+      secure: (loc.protocol === 'https:') ? 1 : 0,
```
Why not `true` and `false` instead of `1` and `0`?

muuki88 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,408 @@
+    );
+
+    if (request.errors && request.errors.length > 0) {
+      console.log('ID WARNING 0x01');
```
Does `prebid.js` not have any logging facilities?

dbemiller replied:
`src/utils` has `logWarn` and `logError` functions

agregorio-improve replied:
Changed the logging to utils.logError.

0 and 1 instead of true or false is a throwback to our internal protocol

muuki88 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,408 @@
+
+    var baseUrl = (requestParameters.secure === 1 ? 'https' : 'http') + '://' + this.CONSTANTS.AD_SERVER_BASE_URL + '/';
+
+    var bidRequestObject;
```
why not initialize here directly?

agregorio-improve replied:
Fixed.

agregorio-improve commented:
All concerns have been addressed and automated checks are passing.  Please approve the pull request.

agregorio-improve commented:
Hi,

I don't see the approval.  Merge is still blocked.  Can you please take a look?

Cheers.
Alfredo

muuki88 commented:
@agregorio-improve I'm not a maintainer or contributor, so I can just give my approval, but not actually merge. @dbemiller or @harpere can do this.

dbemiller commented:
Each PR needs two approvals. @harpere is the first reviewer for this one (see the assignees column). Once they're done, they'll tag someone else on the project to be the second.

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+  }
+
+  $$PREBID_GLOBAL$$.improveDigitalResponse = function(response) {
+    let bidRequests = $$PREBID_GLOBAL$$._bidsRequested.find(bidSet => bidSet.bidderCode === IMPROVE_DIGITAL_BIDDER_CODE);
```
Use `utils.getBidderRequestAllAdUnits` or `util.getBidRequest` here

agregorio-improve replied:
Fixed

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+    };
+
+    let normalizedBids = [];
+    for (var bidCounter = 0; bidCounter < bidRequests.length; bidCounter++) {
```
You can use a https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map over here and all the other places in this file. No need to create an array and push..  your code will be much cleaner.

```
let normalizedBids = bidRequest.map((bidRequest) => {
    // your code
    return result;
});

```

agregorio-improve replied:
Fixed.

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+    }
+
+    if (request && request.requests && request.requests[0]) {
+      for (var requestCounter = 0; requestCounter < request.requests.length; requestCounter++) {
```
Use utils._each instead of all the for loops

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+module.exports = ImproveDigitalAdapter;
+
+function ImproveDigitalAdServerJSClient(endPoint) {
+  this.isA = function (object, _t) {
```
We have `isA`, `isArray` functions in utils.js

agregorio-improve replied:
Fixed

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+    return this.isA(object, 'Array');
+  };
+
+  this.getRandomString = function(prefix) {
```
Check if you can use `utils.generateUUID` or `utils.getUniqueIdentifierStr`

agregorio-improve replied:
Fixed

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+    };
+  };
+
+  this.createGetRequest = function(requestObject, requestParameters, extraRequestParameters) {
```
Where are you using this function ?

agregorio-improve replied:
Fixed.

jaiminpanchal27 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,414 @@
+          if (bidObject.id === bidRequest.bidId) {
+            let bid;
+            if (!bidObject.price || bidObject.price === null) {
+              bid = bidfactory.createBid(CONSTANTS.STATUS.NO_BID);
```
Pass `bidRequest` as second param to `createBid`. Required to match request/response pair

agregorio-improve replied:
Fixed

mkendall07 commented:
@agregorio-improve 
There needs to some updates on this before we can merge. Thanks

agregorio-improve commented:
All comments have been addressed.  Please approve this merge request.

agregorio-improve commented:
Looks like @harpere did not approve my changes despite the fact that I have addressed all the issues specified.  Can someone please remove this block?

agregorio-improve commented:
Thank you @muuki88 and @jaiminpanchal27 but it looks like @harpere is still blocking this.  Anything you can do?

agregorio-improve commented:
Hi,

Can someone unblock this please?

Regards,
Alfredo

Re-lM commented:
Hi all,
What is the status of this one? Can it be merged?
Thanks!

mkendall07 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,386 @@
+const LIB_VERSION_GLOBAL = '3.0.5';
+
+var CONSTANTS = require('src/constants');
+var utils = require('src/utils');
```
we prefer `let` and `const` over `var`

agregorio-improve replied:
Fixed

mkendall07 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,386 @@
+  let LIB_VERSION = LIB_VERSION_GLOBAL;
+
+  // Ad server needs to implement JSONP using this function as the callback
+  let CALLBACK_FUNCTION = '$$PREBID_GLOBAL$$' + '.improveDigitalResponse';
```
FYI we won't allow JSONP callbacks with prebid 1.0.

agregorio-improve replied:
We support CORS and will update our adapter when required by prebid 1.0.

mkendall07 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,386 @@
+          if (bidObject.id === bidRequest.bidId) {
+            let bid;
+            if (!bidObject.price || bidObject.price === null) {
+              bid = bidfactory.createBid(CONSTANTS.STATUS.NO_BID, bidRequest);
```
repeated code blocks should go into a function

agregorio-improve replied:
Fixed.

mkendall07 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -0,0 +1,386 @@
+
+            let nurl = '';
+            if (bidObject.nurl && bidObject.nurl.length > 0) {
+              nurl = '<img src=\"' + bidObject.nurl + '\" width=\"0\" height=\"0\" style=\"display:none\">';
```
prefer template literals over string concatenation. (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)

agregorio-improve replied:
Fixed

mkendall07 commented:
@agregorio-improve @Re-lM  
let a few comments.

agregorio-improve commented:
All outstanding issues have been addressed.  Please provide feedback and/or approve merge.

mkendall07 commented on modules/improvedigitalBidAdapter.js:
```diff
@@ -62,50 +62,48 @@ const ImproveDigitalAdapter = function () {
-              syncString += (syncString === '') ? 'document.writeln(\"' : '';
               let syncInd = syncElement.replace(/\//g, '\\\/');
-              syncString += '<img src=\\\"' + syncInd + '\\\"\/>';
+              syncString = `${syncString}${(syncString === '') ? 'document.writeln(\"' : ''}<img src=\\\"${syncInd}\\\" style=\\\"display:none\\\"\/>`;
```
You don't need to escape `"` anymore
